src/core/io/src/4C_io_input_spec_validators.hpp

```diff
@@ -30,78 +31,202 @@ FOUR_C_NAMESPACE_OPEN
  */
 namespace Core::IO::InputSpecBuilders::Validators
 {
-  /**
-   * A generic validator providing the interface to check an input value against a predicate.
-   * This class is heavily tied to the InputSpec system and is used to validate input values in a
-   * type-erased manner. Usually, there is nothing you should do with this class directly. Use the
-   * various factory functions to create instances of this class that can then be passed to one of
-   * the InputSpecBuilders functions.
-   */
-  template <typename T>
-  class Validator
+  namespace Internal
   {
-    using Predicate = std::function<bool(const T&)>;
-    using Describe = std::function<void(std::ostream&)>;
-    using EmitMetadata = std::function<void(YamlNodeRef)>;
+    template <typename T>
+    struct RangeLike
+    {
+      using value_type = T;
+    };
 
-    struct Concept
+    /**
+     * A type trait that abstracts certain types into a more general form. For example,
+     * a std::vector<T> is abstracted into RangeLike<T> to indicate that a validator may be
+     * applied to all elements of the vector.
+     */
+    template <typename T, typename Enable = void>
+    struct GeneralizedType;
```

Comment:
... and this helper does the recursive conversion of something like `std::vector<std::vector<int>>` into
`RangeLike<RangeLike<int>>`.